### PR TITLE
Tweet about upcoming tweets in 5 minutes

### LINKED ISSUE #63: Send out a tweet 5mins before the stream goes live.

### PR DESCRIPTION
**Pull request by spaantje**
Tweet about the upcoming streams. 

Some things to consider: 
- What text do you want to use for the tweet. Using now: `🔴 A new stream{$twitterHandleIfGiven}is about to start: {$this->stream->title}. Join in now!" . PHP_EOL . $this->stream->url()`
- Do we like the name `announcement_tweeted_at` for the extra database column (& do we actually want an extra column?)

Fixes #63

### CODE REVIEW

File: tests/Feature/Commands/TweetAboutUpcommingStreamsTest.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        $this->assertNotNull($streamToTweet->refresh()->announcement_tweeted_at);
+        $this->assertNull($streamDontTweet6Min->refresh()->announcement_tweeted_at);
+        $this->assertNull($streamDontTweetLive->refresh()->announcement_tweeted_at);
+    }
+
+    /** @test */
+    public function it_only_tweets_streams_that_are_going_live_once(): void
+    {
+        // Arrange
+        Stream::factory()->upcoming()->create(['scheduled_start_time' => now()->addMinutes(5), 'announcement_tweeted_at' => now()]);
```
Comment (christophrumpel) [inline tests/Feature/Commands/TweetAboutUpcommingStreamsTest.php:109]: Please add a factory state method for `announcementTweetSend`, something like this? Or do you have a better name?

**Review by christophrumpel — CHANGES_REQUESTED** [review]
Hey Sonny, I just added a few change requests. Let me know if you have any questions.

File: tests/Feature/Commands/TweetAboutUpcomingStreamsCommandTest.php [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            $this->assertInstanceOf(Carbon::class, $upcomingStreamToTweet->announcement_tweeted_at);
+            $this->assertSame((string) now(), (string) $upcomingStreamToTweet->announcement_tweeted_at);
+        });
+
+        $this->assertNull($upcomingStreamNotToTweet->refresh()->announcement_tweeted_at);
+        $this->assertNull($liveStreamNotToTweet->refresh()->announcement_tweeted_at);
+    }
+
+    /** @test */
+    public function it_does_not_send_a_tweet_if_the_live_tweet_is_sent(): void
```
Comment (spaantje) [inline tests/Feature/Commands/TweetAboutUpcomingStreamsCommandTest.php:113]: @christophrumpel i have added this test (and corresponding check in the job) to make sure we don't tweet about an upcoming stream if the tweet for the "stream is live" is already sent. This should not happen but in the odd case the job is running long or something we should not tweet it out i think?

**Review by christophrumpel — APPROVED** [review]
Looks good, thanks!

File: tests/Feature/Commands/TweetAboutLiveStreamsCommandTest.php [hunk 3]
```diff
@@ -29,6 +30,20 @@ public function it_only_tweets_streams_that_are_live(): void
         $this->twitterFake->assertTweetWasSent();
     }
 
+    /** @test */
+    public function it_does_not_tweet_streams_that_are_live_or_finished(): void
```
Comment (spaantje) [inline tests/Feature/Commands/TweetAboutLiveStreamsCommandTest.php:34]: @christophrumpel Shouldn't this be `it_does_not_tweet_streams_that_are_upcoming_or_finished` ?

File: app/Console/Commands/TweetAboutUpcomingStreamsCommand.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+    protected $description = 'Tweet announcement for upcoming streams.';
+
+    public function handle(): int
+    {
+        $streams = Stream::query()
+            ->approved()
+            ->upcoming()
+            ->where('scheduled_start_time', '<=', now()->addMinutes(5))
+            ->whereNull('upcoming_tweeted_at')
```
Comment (christophrumpel) [inline app/Console/Commands/TweetAboutUpcomingStreamsCommand.php:21]: I decided to change the field name.

File: tests/Feature/Commands/TweetAboutUpcomingStreamsCommandTest.php (hunk 2 above)
Comment (christophrumpel) [reply tests/Feature/Commands/TweetAboutUpcomingStreamsCommandTest.php:127]: yeah good catch 👍

**Comment by christophrumpel** [comment]
Alright, I did some naming refactoring. I'm happy now, you too?

**Comment by spaantje** [comment]
@christophrumpel i'm wondering if we should add a Factory state like `upcomingStreamToTweet` (or something) where we set the `['scheduled_start_time' => now()->addMinutes(5)]` if you want to change the period to anything other than 5 minutes in the future we only have to change the factory and don't have to change all the tests. What do you think ?

**Comment by christophrumpel** [comment]
Good idea, I thought about that too, but was unsure how to call it. `upcomingStreamToTweet´ is a little misleading I think. What about `startsInUpcomingTweetRange`?

**Comment by spaantje** [comment]
What about `withinUpcomingTweetRange` ?

**Comment by christophrumpel** [comment]
yeah already good but `what` is in that range. I want to be connected somehow to the start time so it is more obvious when you read it. Maybe `startWithinUpcomingTweetRange`?

**Comment by spaantje** [comment]
I have changed it with the name `startWithinUpcomingTweetRange`. I don't really love it yet. But don't know a better one. I think the PR is done.

**Comment by christophrumpel** [comment]
Thanks for your work on this PR

**Comment by spaantje** [comment]
Your welcome!

**Pull request merged**